lib/Conversion/ArcToLLVM/LowerArcToLLVM.cpp

```diff
@@ -285,6 +291,239 @@ struct ReplaceOpWithInputPattern : public OpConversionPattern<OpTy> {
 
 } // namespace
 
+//===----------------------------------------------------------------------===//
+// Simulation Orchestration Lowering Patterns
+//===----------------------------------------------------------------------===//
+
+static Value createPtrToPortState(ConversionPatternRewriter &rewriter,
+                                  Location loc, Value state,
+                                  const StateInfo *port) {
+  MLIRContext *ctx = rewriter.getContext();
+  return rewriter.create<LLVM::GEPOp>(loc, LLVM::LLVMPointerType::get(ctx),
+                                      IntegerType::get(ctx, 8), state,
+                                      LLVM::GEPArg(port->offset));
+}
+
+namespace {
+
+struct SimInstantiateLowering
+    : public OpConversionPattern<arc::SimInstantiate> {
+  SimInstantiateLowering(const TypeConverter &typeConverter,
+                         MLIRContext *context, ArrayRef<ModelInfo> modelInfo)
+      : OpConversionPattern(typeConverter, context), modelInfo(modelInfo) {}
+
+  LogicalResult
+  matchAndRewrite(arc::SimInstantiate op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    const ModelInfo *model = llvm::find_if(
+        modelInfo, [&](auto x) { return x.name == op.getType().getModel(); });
+    if (model == modelInfo.end()) {
+      op.emitError("model not found");
+      return failure();
+    }
+
+    ConversionPatternRewriter::InsertionGuard guard(rewriter);
+
+    Location loc = op.getLoc();
+    Value numStateBytes = rewriter.create<LLVM::ConstantOp>(
+        loc, rewriter.getIntegerType(sizeof(size_t) * 8), model->numStateBytes);
+    Value allocated = rewriter.replaceOpWithNewOp<LLVM::AllocaOp>(
+        op, LLVM::LLVMPointerType::get(getContext()), rewriter.getI8Type(),
+        numStateBytes);
```

Comment:
Are you sure you want to allocate the whole simulation state on the stack? I assume you want to call `malloc` here?

Reply:
You are right, in particular it gets bad if one tries to return the model from a function. What would the freeing behavior be? Should I introduce a free operation (easy, but may leak memory), or should I scope the existence of the model in a region, and have `arc.sim.instantiate` be a region operation?

Reply:
Is there a foreseeable need to arbitrarily pass around model instances (pass to functions, return from functions, etc.)? If not, starting with constraining it to a region would be better (generalising later on is typically better than restricting).

Reply:
In order to call malloc, I need to create a `size_t` to pass the size I want to libc. The only way to know the size of `size_t` is to introduce a dependency on clang because it is part of the C ABI which LLVM is not aware of. Is it worth it?

Reply:
Another way is to confine the value in a region and still use alloca. It may be an issue on large model but I also assume increasing stack size is a possibility?

Reply:
AFAIK the memref dialect also has lowering options to use malloc and uses an `index` type there, so I guess MLIR handles this type the same way?

Reply:
Like in the memref dialect it could also be a pass option whether alloca or malloc should be used.